<h3>We will now consider the Boston housing data set, from the ISLP library.

In [2]:
from ISLP import load_data
import numpy as np
import pandas as pd

Boston = load_data('Boston')
Boston.head()

,crim,zn,indus,chas,nox,rm,age,dis,rad,tax,ptratio,lstat,medv
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296,15.3,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242,17.8,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242,17.8,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222,18.7,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222,18.7,5.33,36.2


### (a) Based on this data set, provide an estimate for the population mean of medv. Call this estimate $\hat{\mu}$.

In [3]:
u_hat = np.mean(Boston['medv'])
u_hat

22.532806324110677

### (b) Provide an estimate of the standard error of $\hat{\mu}$. Interpret this result.
Hint: We can compute the standard error of the sample mean by dividing the sample standard deviation by the square root of the number of observations.

In [4]:
medv = Boston['medv']

n = len(medv)
s = np.std(medv, ddof=1)
se_u = s / np.sqrt(n)
se_u

0.4088611474975351

The estimated mean home value is \( $\hat{\mu}$ = 22.53 \), with a standard error of \( 0.41 \).  
if we repeatedly sampled, the sample mean of `medv` would typically vary by about \$410 around the true population mean.

### (c) Now estimate the standard error of $\hat{\mu}$ using the bootstrap. How does this compare to your answer from (b)?

In [5]:
medv = Boston['medv'].values
n = len(medv)

rng = np.random.default_rng(42)
B = 10000
boot_means = np.empty(B)
for b in range(B):
    idx = rng.integers(0, n, size=n)
    boot_means[b] = medv[idx].mean()

boot_se = boot_means.std(ddof=1)
boot_se

0.4092512845531801

### (d) Based on your bootstrap estimate from (c), provide a 95 % confidence interval for the mean of medv. Compare it to the results obtained by using Boston['medv'].std() and the two standard error rule (3.9).
Hint: You can approximate a 95 % confidence interval using the formula
$
\left[ \hat{\mu} - 2\,\mathrm{SE}(\hat{\mu}), \; \hat{\mu} + 2\,\mathrm{SE}(\hat{\mu}) \right]
$


In [6]:
CI_95_low =  u_hat - 2*boot_se
CI_95_high = u_hat + 2*boot_se

ci = [CI_95_low,CI_95_high]

print(ci)

[21.714303755004316, 23.351308893217038]


use bootstrap estimate: u hat is 22.53, bootstrap SE mean is 0.41, 95% confidence interval is (21.71, 23.35)

In [7]:
std = Boston['medv'].std()
n = Boston['medv'].count()
# standard error u hat
se_u_hat = std/(n**(1/2))
print(std)
print(n)
print(se_u_hat)
# use 2 standard error rule
CI_2ser_low =  u_hat - 2*se_u_hat
CI_2ser_high = u_hat + 2*se_u_hat

ci_2ser = [CI_2ser_low,CI_2ser_high]

print(ci_2ser)

9.19710408737982
506
0.4088611474975351
[21.715084029115605, 23.35052861910575]


use 2 standard error rule: u hat is 22.53, standard error mean is 0.41, 95% confidence interval is (21.72,23.35)
bootstrap standard error ci and 2 standard error ci are identical, because there are 506 observations so that sampling distribution is close enough to normal

### (e) Based on this data set, provide an estimate, $\hat{\mu}_{\text{med}}$, for the median value of medv in the population.

In [8]:
u_med_hat = np.median(Boston['medv'])
u_med_hat

21.2

### (f) We now would like to estimate the standard error of $\hat{\mu}_{\text{med}}$. Unfortunately, there is no simple formula for computing the standard error of the median. Instead, estimate the standard error of the median using the bootstrap. Comment on your findings.

In [9]:
medv = Boston['medv'].values
n = len(medv)

rng = np.random.default_rng(42)
B = 10000
boot_medians = np.empty(B)

for b in range(B):
    sample = rng.choice(medv, size=n, replace=True)
    boot_medians[b] = np.median(sample)

boot_se_med = boot_medians.std(ddof=1)
boot_se_med

0.38034789920208906

we resample many times then the estimated standard error of the median is 0.38, less than se of the mean (0.41). because median is less sensitive to outlier value.

### (g) Based on this data set, provide an estimate for the tenth percentile of medv in Boston census tracts. Call this quantity $\hat{\mu}_{\text{0.1}}$. (You can use the np.percentile() function.)

In [10]:
medv = Boston['medv'].values

u_01_hat = np.percentile(medv, 10)
u_01_hat

12.75

### (h) Use the bootstrap to estimate the standard error of μˆ0.1 percentile() . Comment on your findings.

In [11]:
medv = Boston['medv'].values
n = len(medv)

rng = np.random.default_rng(42)
B = 10000
boot_p10 = np.empty(B)

for b in range(B):
    idx = rng.integers(0, n, size=n)
    sample = medv[idx]
    boot_p10[b] = np.percentile(sample, 10)

boot_se_p10 = boot_p10.std(ddof=1)
boot_est_p10 = np.percentile(medv, 10)

boot_est_p10, boot_se_p10

(12.75, 0.5031575068978646)

the estimated se of 10th percentile is 12.75. bootstrap se of 10th percentile is 0.5. it is higher than se of mean(0.41) and median(0.38). because 10th percentile depend on fewer effective observations